learn/building_programs/distributing.md

```diff
@@ -0,0 +1,135 @@
+---
+layout: book
+title: Distributing your programs
+permalink: /learn/building_programs/distributing
+---
+
+When you distribute your programs, there are a number of options you can
+choose from:
+
+1. Distribute the entire source code
+2. Distribute a pre-built executable program
+3. Distribute static or dynamic libraries that people can use
+
+ad 1.
+
+By far the simplest - for you as a programmer - is this one: you leave it
+up to the user to build it on their own machine. Unfortunately, that
+means you will have to have a user-friendly build system in place and
+the user will have to have access to suitable compilers. For build systems:
+see the previous section.
+
+ad 2.
+
+A pre-built program that does not need to be customised, other than via its
+input, will still need to come with the various run-time libraries and will
+be specific to the operating system/environment it was built for.
+
+The set of run-time libraries differs per operating system and compiler version.
+For a freely available compiler like gfortran, the easiest thing is to ask the
+user to install that compiler on their system. In the case of Windows: the Cygwin
+environment may be called for.
+
+Alternatively, you can supply copies of the run-time libraries together with your
+program. Put them in the directory where they can be found at run-time.
+
+Note: On Windows, the Intel Fortran comes with a set of _redistributable_ libraries.
+These will need to be made available.
+
+In general: use a tool like "ldd" or "dependency walker" to find out what
+libraries are required and consult the documentation of the compiler.
+
+If your program does allow customisation, consider using dynamic libraries for this.
+More is said about this below.
+
+ad 3.
+
+This option is a combination of the first two options. It does put some burden on
+the user, as they must create a main program that calls your routines in the
+proper way, but they do not need to know much about the build system you used.
+You will have to deal with the run-time libraries, though.
+
+If you choose this option, besides the compiled libraries, you will also need to
+supply the module intermediate files. These files are compiler-specific, but so are
+the static libraries you build.
+
+## Distributing the tabulation program
+As shown above, the tabulation program can be built with the user-defined function
+in a dynamic library. This enables you to:
+
+* Ship the executable (with the appropriate run-time libraries)
+* Provide a skeleton version of the module, something like:
+
+```fortran
+        module user_functions
+            implicit none
+        contains
+
+        real function f( x )
+        !DEC$ ATTRIBUTES DLLEXPORT :: f
+            real, intent(in) :: x
+
+            ... TO BE FILLED IN ...
+
+        end function f
+        end module user_functions
+```
+
+* Provide a basic build script with a command like:
+
+```shell
+        gfortran -o function.dll function.f90 -shared
+```
+
+        or:

```

Comment:
For some reason this 'or' is being rendered in a code block, but I can't work out why?

Reply:
Okay, removing the indent fixes it.

```suggestion
or:
```